### PR TITLE
WS master should send plugin IDs instead of plugins metas to unified broker

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Makes ws master send plugin FQNs to broker instead of downloaded meta.yamls.

Draft PR until https://github.com/eclipse/che-plugin-broker/pull/40 is merged.

### What issues does this PR fix or reference?
https://github.com/redhat-developer/rh-che/issues/1307; related to https://github.com/eclipse/che/issues/12908

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: I know that you took these cases from another test class. But could you rework this test to test with data provider? It would looks way simpler and could include all the non exception cases like what you have in `shouldReturnEmptyListWhenNoPluginsOrEditors`.
Apart from making code more concise and simplify its support you could split cases where different bugs would not fail the same test case as it is now with this test. 
E.g. 
- case with registry
- case with registry that has path /plugins
- case with registry that has path not equal /plugins
- case with registry that has multi-level path
- case without registry
- case with null editor
- case with null plugins
- case with multiple plugins
- etc

To simplify matching of results you can use `assertEqualsnoOrder(actual.toArray(), expected.toArray())` (please excuse not precise names of methods if any)

**Review by garagatyi — COMMENTED** [review]
Looks good!

**Comment by garagatyi** [comment]
Please take into account that after merging this PR we will need to adapt RhChe https://github.com/redhat-developer/rh-che/pull/1192/files#diff-d1c7cee2e326bd168cb76031bef4ef0cR197

**Comment by amisevsk** [comment]
Marked as ready for review since it works with locally built plugin-broker, but should not be released until https://github.com/eclipse/che-plugin-broker/pull/46 is merged and a new image is available.

**Comment by amisevsk** [comment]
For testing, images `amisevsk/che-unified-plugin-broker:v0.15.2` and `amisevsk/che-init-plugin-broker:v0.15.2` can be used.

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: The full expected message is `String.format("Invalid Che tooling plugins configuration: plugin %s:%s is duplicated", ...)` and I think it makes sense to check full message to make sure that message is formatted correctly (message contains `ID:VERSION` but not other values). I started to ask it after I saw that Che Server respond with errors where messages are wrong formatted, and it's pretty easy to make mistake here especially during refactoring.
WDYT?
It makes less sense to check the full message that is not formatted but still makes sense. It's like double fields for putting a password (password, confirm password). Also, if the message becomes formatted - then the author should change tests accordingly, otherwise it would be changed at all.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java]: ` *, *` makes sure that empty spaces will be removed between plugins but it does not remove empty spaces before the first plugin and after the last plugin. Maybe it would be safer to split by `,` and then trim split values.
```java
  @Test(
      expectedExceptions = InfrastructureException.class,
      expectedExceptionsMessageRegExp = "Invalid.*duplicated")
  public void shouldDetectDuplicatedPluginsIfTheyArePrefixedSuffixedWithEmptySpaces() throws Exception {
    Map<String, String> attributes =
        createAttributes(
            "",
            " " + formatPlugin("http://testregistry1:8080", "testplugin", "1.0"),
            formatPlugin("http://testregistry2:8080", "testplugin", "1.0") + " ");

    parser.parsePlugins(attributes);
  }
```

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:67]: Please add `toString` as well. It may be useful in tests

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:62]: `hashCode` might look simpler like
```java
  @Override
  public int hashCode() {
    return Objects.hash(getRegistry(), getId(), getVersion());
  }
```
But I'm OK with current way as well.

**Review by sleshchenko — APPROVED** [review]
Overall looks good to me.

Please take a look my inline comments

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java
Comment (amisevsk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java]: Good point, moved to package private.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java
Comment (amisevsk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java]: Updated java doc and removed brokersResult -- good catch that must be leftover from the merging of brokers.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java
Comment (amisevsk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java]: Tha's what I get for writing code to fit what's already there :)

Removed annotations.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java
Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java:47]: Added docs

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java
Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: Yeah it makes sense, I usually default to keeping the match as general as possible (only to ensure exception is being thrown for the right reason) but validating that plugin id and version are included makes a lot of sense.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java
Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java]: Good catch -- done. And thanks for the test case, I've added it to the suite.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java
Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:67]: Done.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java
Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:62]: The pitfalls of auto-generated code. Fixed.

**Comment by amisevsk** [comment]
ci-test

**Comment by amisevsk** [comment]
Waiting on a release of Che plugin broker to mirror the necessary changes.

**Do not merge**

**Comment by garagatyi** [comment]
@amisevsk I see that there is a release with necessary changes but it is done manually instead of by CI and container image is not pushed to Docker hub. Do you plan to push them? Do you need help with that?

**Review by garagatyi — APPROVED** [review]

**Comment by amisevsk** [comment]
@garagatyi Thanks for the help, I couldn't figure out the CI backing plugin-broker releases. 

Bumped broker version in che.properties, squashed all commits, and rebased on current master. Tested locally and everything works with FQN changes.

**Pull request merged**
